framework/Floats.cpp

```diff
@@ -0,0 +1,194 @@
+
+static_assert(sizeof(Float64) == sizeof(double), "Float64 structure is not of the correct size");
+static_assert(FLOAT64_NAN_EXPONENT == FLOAT64_EXPONENT_MASK, "Float64::NaNs have all exponent bits set");
+static_assert(FLOAT64_INFINITY_EXPONENT == FLOAT64_EXPONENT_MASK, "Float64::Inf has all exponent bits set");
+static_assert(MASK(FLOAT64_EXPONENT_BITS) == FLOAT64_EXPONENT_MASK, "Float64 exponent mask has different size than the field");
+static_assert(MASK(FLOAT64_MANTISSA_BITS) == FLOAT64_MANTISSA_MASK, "Float64 mantissa mask has different size than the field");
+static_assert(QUIET(FLOAT64) == FLOAT64_MANTISSA_QUIET_NAN_MASK, "Quiet bit is MSB of the mantissa");
+static_assert(FLOAT64_SIGN_BITS + FLOAT64_EXPONENT_BITS + FLOAT64_MANTISSA_BITS == 64, "Bitfields sums to type size");
+
+static_assert(sizeof(double) < sizeof(long double), "Compiler does not support long double");
+static_assert(sizeof(Float80) == sizeof(long double), "Float80 structure is not of the correct size");
+static_assert(FLOAT80_NAN_EXPONENT == FLOAT80_EXPONENT_MASK, "All Float80::NaNs have all exponent bits set");
+static_assert(FLOAT80_INFINITY_EXPONENT == FLOAT80_EXPONENT_MASK, "Float80::Inf has all exponent bits set");
+static_assert(MASK(FLOAT80_EXPONENT_BITS) == FLOAT80_EXPONENT_MASK, "Float80 exponent mask has different size than the field");
+static_assert(MASK(FLOAT80_MANTISSA_BITS) == FLOAT80_MANTISSA_MASK, "Float80 mask has different size than mantissa+jbit");
+static_assert(QUIET(FLOAT80) == FLOAT80_MANTISSA_QUIET_NAN_MASK, "Quiet bit is MSB of the mantissa");
+static_assert(FLOAT80_SIGN_BITS + FLOAT80_EXPONENT_BITS + FLOAT80_JBIT_BITS + FLOAT80_MANTISSA_BITS == 80, "Bitfields sums to type size");
+
+/** @} */
+
+/**
+ * @brief C function definitions
+ * @{
+ */
+extern "C" {
+
+Float16 new_random_float16()
+{
+    Float16 f;
+    f.sign = random32();
+    f.exponent = random32();
+    f.mantissa = set_random_bits(random32() % (FP16_MANTISSA_BITS + 1), FP16_MANTISSA_BITS);
+
+    return f;
+}
+
+BFloat16 new_random_bfloat16()
+{
+    BFloat16 f;
+    f.sign = random32();
+    f.exponent = random32();
+    f.mantissa = set_random_bits(random32() % (BFLT16_MANTISSA_BITS + 1), BFLT16_MANTISSA_BITS);
+
+    return f;
+}
+
+Float32 new_random_float32()
+{
+    Float32 f;
+    f.sign = random32();
+    f.exponent = random32();
+    f.mantissa = set_random_bits(random32() % (FLOAT32_MANTISSA_BITS + 1), FLOAT32_MANTISSA_BITS);
+
+    return f;
+}
+
+Float64 new_random_float64()
+{
+    Float64 f;
+    f.sign = random32();
+    f.exponent = random32();
+    f.mantissa = set_random_bits(random32() % (FLOAT64_MANTISSA_BITS + 1), FLOAT64_MANTISSA_BITS);
+
+    return f;
+}
+
+Float80 new_random_float80()
+{
+    Float80 f;
+    f.sign = random32();
+    f.exponent = random32();
+    f.jbit = 1;
+    f.mantissa = set_random_bits(random32() % (FLOAT80_MANTISSA_BITS + 1), FLOAT80_MANTISSA_BITS);
```

Comment:
This is the only use of `FLOAT80_MANTISSA_BITS`. But because we do set `jbit` above, the mask without the MSB is *correct*.

What's not correct is the use of only 32 bits of randomness for a 64-bit field. Ditto for the `double` case above. But pre-existing.